File: backend/tour-service/src/main/java/com/project/tour/service/port/PortService.java
package com.project.tour.service.port;

import com.project.common.dto.location.AddressResponse;
import com.project.common.service.location.GeocodingService;
import com.project.tour.dto.port.CreatePortRequest;
import com.project.tour.dto.port.PortResponse;
import com.project.tour.dto.port.UpdatePortRequest;
import com.project.tour.exception.AppException;
import com.project.tour.mapper.port.PortMapper;
import com.project.tour.model.Port;
import com.project.tour.model.enums.PortStatus;
import com.project.tour.repository.PortRepository;

import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@Transactional
public class PortService {

    private final PortRepository portRepository;
    private final PortMapper portMapper;
    private final GeocodingService geocodingService;

    public PortService(
            PortRepository portRepository,
            PortMapper portMapper,
            GeocodingService geocodingService) {

        this.portRepository = portRepository;
        this.portMapper = portMapper;
        this.geocodingService = geocodingService;
    }

    // CREATE
    public PortResponse createPort(CreatePortRequest request) {

        Port port = portMapper.toEntity(request);

        AddressResponse address = geocodingService.getAddress(
                request.getLatitude(),
                request.getLongitude());

        if (address != null) {
            port.setAddress(address.getFullAddress());
            port.setCity(address.getCity());
            port.setCountry(address.getCountry());
        }

        Port savedPort = portRepository.save(port);

        return portMapper.toResponse(savedPort);
    }

    // GET ALL
    @Transactional(readOnly = true)
    public List<PortResponse> getAllPorts() {

        return portRepository
                .findAll(Sort.by(Sort.Direction.ASC, "name"))
                .stream()
                .map(portMapper::toResponse)
                .toList();
    }

    // GET ACTIVE
    @Transactional(readOnly = true)
    public List<PortResponse> getActivePorts() {

        return portRepository
                .findAllByStatusOrderByNameAsc(PortStatus.ACTIVE)
                .stream()
                .map(portMapper::toResponse)
                .toList();
    }

    // GET BY ID
    @Transactional(readOnly = true)
    public PortResponse getPortById(UUID id) {

        Port port = findPortById(id);

        return portMapper.toResponse(port);
    }

    // UPDATE
    public PortResponse updatePort(
            UUID id,
            UpdatePortRequest request) {

        Port port = findPortById(id);

        portMapper.updateEntity(port, request);

        AddressResponse address = geocodingService.getAddress(
                request.getLatitude(),
                request.getLongitude());

        if (address != null) {
            port.setAddress(address.getFullAddress());
            port.setCity(address.getCity());
            port.setCountry(address.getCountry());
        }

        Port updatedPort = portRepository.save(port);

        return portMapper.toResponse(updatedPort);
    }

    // SOFT DELETE
    public void deactivatePort(UUID id) {

        Port port = findPortById(id);

        port.setStatus(PortStatus.INACTIVE);

        portRepository.save(port);
    }

    // FIND ENTITY
    private Port findPortById(UUID id) {

        return portRepository.findById(id)
                .orElseThrow(() -> new AppException(
                        "Port not found",
                        HttpStatus.NOT_FOUND));
    }
}

File: frontend-web/src/modules/admin/services/portService.js
// src/modules/admin/services/portService.js

import api from "../../../api/axios";

const PORT_BASE_URL = "/admin/ports";

const portService = {
  // =====================================================
  // GET ALL PORTS
  // GET /api/admin/ports
  // =====================================================

  async getPorts(activeOnly = false) {
    const response = await api.get(PORT_BASE_URL, {
      params: {
        activeOnly,
      },
    });

    return response.data;
  },

  // =====================================================
  // GET PORT BY ID
  // GET /api/admin/ports/{id}
  // =====================================================

  async getPortById(id) {
    const response = await api.get(`${PORT_BASE_URL}/${id}`);

    return response.data;
  },

  // =====================================================
  // CREATE PORT
  // POST /api/admin/ports
  // =====================================================

  async createPort(data) {
    const response = await api.post(PORT_BASE_URL, data);

    return response.data;
  },

  // =====================================================
  // UPDATE PORT
  // PATCH /api/admin/ports/{id}
  // =====================================================

  async updatePort(id, data) {
    const response = await api.patch(`${PORT_BASE_URL}/${id}`, data);

    return response.data;
  },

  // =====================================================
  // DELETE / DEACTIVATE PORT
  // DELETE /api/admin/ports/{id}
  // =====================================================

  async deactivatePort(id) {
    await api.delete(`${PORT_BASE_URL}/${id}`);
  },
};

export default portService;
